### PR TITLE
Improve performance of sloth and team stats for frontend

### PR DESCRIPTION
**Pull request by Astraeus-**
Improves performance significantly of loading statistics for sloths and teams on the front-end pages. This is likely one of our CPU time culprits...

In my dev environment `view/team` and `user/view` pages could take up to 2s to complete for some vs <1s now.

### CODE REVIEW

**Comment by Astraeus-** [comment]
Below some of the measurement results that I got in my local testing. I tried to take some varied sample data with old team vs new team and sloths with varied number of matches.

TEAMS
Team bye:
	Old: 2.45s
	New: 0.56s

Team DOF
	Old: 0.85s
	New: 0.65s

Team callphone
	Old: 0.46s
	New: 0.41s

SLOTHS
Astraeus: id 39
	Old: 0.76s
	New: 0.22s

Ruini: id 125
	Old: 1.07s
	New: 0.31s

Omegakim: id 9237
	Old: 2.7s
	New: 0.15s

File: plugins/rikki/loungestatistics/classes/statistics/Statistics.php [hunk 1]
```diff
@@ -207,22 +207,21 @@ public static function calculateHeroStatisticsForTeam($team, $season)
 
     public static function calculateMapStatisticsForSloth($gameParticipations, $season)
     {
-        $games = [];
+        $participations = [];
```
[the remark this answers is not in the document]
Comment (Astraeus-) [reply plugins/rikki/loungestatistics/classes/statistics/Statistics.php:210]: The input parameter of that method already has that name, would you prefer they're swapped around?

File: plugins/rikki/loungestatistics/classes/statistics/Statistics.php [hunk 2]
```diff
@@ -287,4 +286,39 @@ public static function analyzeGamesForMapStats($games, $teams) {
...
+            if ($map_array['picks_by'] + $map_array['picks_vs'] != 0) {
+                $map_array['winrate'] = round($map_array['winrate']/(($map_array['picks_by'] + $map_array['picks_vs'])*0.01),2);
+                $map_array['winrate'] .= '%';
+            } else {
+                $map_array['winrate'] = '-';
+            }
+            $mapArray[$key] = $map_array;
+        }
+        return $mapArray;
+    }
```
[the remark this answers is not in the document]
Comment (Astraeus-) [reply plugins/rikki/loungestatistics/classes/statistics/Statistics.php:323]: I'm not quite sure what you mean by using a callback to exchange the one line. I know there's a lot of duplication, but the old one iterates over games, whereas this one iterates over participations.

File: plugins/rikki/loungestatistics/classes/statistics/Statistics.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (Fabian-Sommer) [reply plugins/rikki/loungestatistics/classes/statistics/Statistics.php:210]: Both their names should reflect the difference between them - what is participations right now should be gameParticipationsInSeason or so

File: plugins/rikki/loungestatistics/classes/statistics/Statistics.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (Fabian-Sommer) [reply plugins/rikki/loungestatistics/classes/statistics/Statistics.php:323]: As far as I can see this is different exactly in line 303 modulo renaming variables ($participations vs $games, $participation->game vs $game).

Thinking about how I would refactor this, I arrive at pretty much the previous state. Please test if this change actually improves performance.

[the remark this answers is not in the document]
Comment (Astraeus-) [reply plugins/rikki/loungestatistics/classes/statistics/Statistics.php:323]: I added my dev benchmarks above. I can get data purely for sloth maps if you want those. 

Currently for sloths we get the game info from their participations. It creates a lot of overhead to fetch all the gameParticipations for each game to determine who picked the map, when there's already a participation to read it off of.

[the remark this answers is not in the document]
Comment (Fabian-Sommer) [reply plugins/rikki/loungestatistics/classes/statistics/Statistics.php:323]: Still no data on how this change affects runtime. Please remove this, or if it has significant impact, refactor. Refactoring options include a function parameter to evaluate the one line that is different, or passing a participation that may be null and using it if it is available.

File: plugins/rikki/heroeslounge/models/GameParticipation.php [hunk 3]
```diff
@@ -57,4 +57,9 @@ public function getOrderedTalentsAttribute($value)
     {
         return $this->talents()->orderBy('talent_tier', 'asc')->get();
     }
+
+    public function isInSecondPickTeam()
+    {
+        return in_array($this->draft_order, [2, 3, 6, 7, 10]);
```
Comment (Fabian-Sommer) [inline plugins/rikki/heroeslounge/models/GameParticipation.php:63]: If you are optimizing, this has a high overhead and could be replaced by draft_order == 2 || ... == 3 || ...

**Comment by Astraeus-** [comment]
Combined the analyzeMap methods back into one, and made a split dependent on if there are also gameParticipations provided.

Below the difference in performance for `analyzeGamesForMapStats` method. `old` is before this PR and `refactor` is from current pr state. The value is the best time of 3 attempts for each type. 

I believe `all time` data, clearly shows that there was indeed room for a lot of performance gain and the initial profile isn't negligible either.

```
Initial profile page / All time

Astraeus: id 39
Old: 0.381s / 8.339s
Refactor: 0.003s / 0.005s

Ruini: id 125
Old: 0.553s / 29.214s
Refactor: 0.002s / 0.012s

Omegakim: id 9237
Old: 1.862s / 1.904s
Refactor: 0.002s / 0.002s
```

**Pull request merged**